homeassistant/components/smartthings/__init__.py

```diff
@@ -100,16 +103,44 @@ async def _refresh_token() -> str:
 
     client.refresh_token_function = _refresh_token
 
+    def _handle_new_subscription_url(url: str) -> None:
+        """Handle a new subscription URL."""
+        hass.config_entries.async_update_entry(
+            entry, data={**entry.data, CONF_SUBSCRIPTION_URL: url}
+        )
+
+    client.refresh_subscription_url_function = _handle_new_subscription_url
+
+    async def subscribe() -> None:
+        """Subscribe to the SmartThings event stream."""
+        try:
+            await client.subscribe(
+                entry.data[CONF_LOCATION_ID],
+                entry.data[CONF_TOKEN][CONF_INSTALLED_APP_ID],
+                entry.data.get(CONF_SUBSCRIPTION_URL),
+            )
+        except (SmartThingsSinkError, SmartThingsForbiddenError) as err:
+            _LOGGER.error("Error subscribing: %s", err)
+            await asyncio.sleep(300)
```

Comment:
Should we add a random extra time to avoid a retry storm?

Reply:
One thing that irritates me with this approach is that we can't properly show the user that the integration couldn't set up. Because this is coming from a background task, it doesn't block startup either.